### PR TITLE
[Accessibility] Associate the shipping calculator button to its form in Cart page

### LINKED ISSUE #51898: Accessibility: Button missing aria-expanded
**Prerequisites (mark completed items with an [x]):**
- [x] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [x] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

### Issue Description:

The screen reader does not announce the state of the <button> element that opens and closes the form to change the shipping address. Screen reader users also do not know if triggering the button has successfully opened or closed the form.

Image

### Component or Page: 

Shortcode: [woocommerce_cart] & Cart Classic Block

### Name affected element:

Change address button

### Relevant code snippets:

`<a href="#" class="shipping-calculator-button">Change address</a>`

### Recommended fix:

Use the aria-expanded attribute to indicate the current state of the button. Set its initial value to ""false"" and toggle it between ""false"" and ""true"" with JavaScript. When implemented correctly, the screen reader will announce changes between expanded and collapsed states.
 
 For more information see https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-expanded

### PR DESCRIPTION
**Pull request by Manussakis**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #51898 .

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. On admin, go to WooCommerce > Settings > Shipping > Shipping settings.
2. Check the "Enable the shipping calculator on the cart page" if it's not checked yet.
3. Go to WooCommerce > Settings > Shipping > Shipping zones.
4. Add a new zone.
5. On the Shipping zones page, add a free Shipping method.
6. On the frontend, add a product to the cart and go to the Cart page.
7. Turn on the screen reader.
8. On the Cart totals section, click on the "Change address" button.
Screenshot 2024-10-17 at 18 02 13
10. Verify that the screen reader announces whether the form is expanded or collapsed.
11. Turn off the screen reader.
12. Go back to the Shipping zones page on the admin and disable the Free Shipping method.
image
13. Save the changes.
14. Go back to the Cart page.
15. Turn on the screen reader.
16. Perform the same tests as before, the only difference is that the button should read something like "Enter address to check delivery options".
Screenshot 2024-10-17 at 18 06 50

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/client/legacy/js/frontend/cart.js [hunk 1]
```diff
@@ -268,8 +268,19 @@ jQuery( function ( $ ) {
...
 		 */
-		toggle_shipping: function () {
-			$( '.shipping-calculator-form' ).slideToggle( 'slow' );
+		toggle_shipping: function ( event ) {
+			var $target = $( event.currentTarget );
+
+			$( '.shipping-calculator-form' ).slideToggle( 'slow', function () {
+				var self = this;
+
+				setTimeout( function () {
```
Comment (Manussakis) [inline plugins/woocommerce/client/legacy/js/frontend/cart.js:277]: When the jQuery `slideToggle` callback runs, the HTML attributes of the `.shipping-calculator-form` element are not 100% updated. The `setTimeout` function ensures that all DOM manipulation made by the slide effect is done, so it's safe to check whether the form is visible or not.

File: plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/totals/shipping/index.tsx [hunk 2]
```diff
@@ -96,6 +96,7 @@ export const TotalsShipping = ( {
 		) : (
 			totalShippingValue
 		);
+	const shippingCalculatorID = 'shipping-calculator-form-wrapper';
```
Comment (Manussakis) [inline plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/totals/shipping/index.tsx:99]: The idea of creating the `shippingCalculatorID` prop for the `<ShippingAddress />`, `<ShippingPlaceholder />`, and `<CalculatorButton />` components is that the `shippingCalculatorID` const can be passed in to all of them from the same parent component, which ensures a single source of truth.

**Comment by Manussakis** [comment]
@dkotter This task is ready for code review.
Thanks!

**Review by dkotter — APPROVED** [review]
@ankitguptaindia @qasumitbagthariya This is ready for QA

**Comment by qasumitbagthariya** [comment]
### QA/Test Report ✅ 
-------------

I have checked this PR in the `Manussakis:fix/51898` branch and changes working as expected.

- verify with Enable and Disable "Free shipping", screen reader announce works as per PR testing notes  ✅ 


https://github.com/user-attachments/assets/26536af6-dfd1-414d-b248-32a0156674ac


https://github.com/user-attachments/assets/e278a823-640e-45e1-b2a2-806cfd957be4



**Testing Environment** 

<Details>

- WordPress: 6.6
- Theme: TT4 v1.2
- WooCommerce - WooCommerce - Dev version with this fix specific branch.
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Git Branch: Manussakis:fix/51898

</Details>

@Manussakis  please remove [WIP] from the title and take this PR out of draft.

**Comment by jeffpaul** [comment]
@mikejolley this PR is ready for your review

**Review by mikejolley — APPROVED** [review]
Thanks, changes look good and test as intended. This can be merged.

**Pull request merged**